### PR TITLE
Move fading to effect and add collision fading support.

### PR DESCRIPTION
**Pull request by Malcolmnixon**
This PR moves the fade logic to a common effect accessible by all nodes - including with the ability to fade to colors.
Additionally it adds collision-based fading to prevent looking through walls.

### CODE REVIEW

**Comment by DigitalN8m4r3** [comment]
lgtm, thanks again malcolm for working this out!

**Comment by Malcolmnixon** [comment]
Added a basic damage area to the basic movement scene.

https://github.com/GodotVR/godot-xr-tools/assets/1863707/74977229-2e61-495f-9bbc-7100c101e8e5

File: addons/godot-xr-tools/effects/fade.gdshader [hunk 1]
```diff
@@ -1,13 +1,13 @@
 shader_type spatial;
 render_mode depth_test_disabled, skip_vertex_transform, unshaded, cull_disabled;
 
-uniform float alpha = 0.0;
+uniform vec4 albedo : source_color;
 
 void vertex() {
 	POSITION = vec4(VERTEX.x, -VERTEX.y, 0.0, 1.0);
```
Comment (BastiaanOlij) [inline addons/godot-xr-tools/effects/fade.gdshader:7]: Something we need to change here that I didn't realise until later is that this needs to be `POSITION = vec4(VERTEX.x, -VERTEX.y, -1.0, 1.0);` when using OpenGL (compatibility). The code as it stands right now is correct for Vulkan.

Once the reverse-Z PR is merged in 4.3, we need to change this again to `POSITION = vec4(VERTEX.x, -VERTEX.y, 1.0, 1.0);`

File: addons/godot-xr-tools/effects/fade.gdshader [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 }
 
 void fragment() {
-	ALBEDO = vec3(0.0, 0.0, 0.0);
-	ALPHA = alpha;
+	ALBEDO = albedo.rgb;
+	ALPHA = albedo.a;
 }
```
Comment (BastiaanOlij) [inline addons/godot-xr-tools/effects/fade.gdshader:13]: Nitpicking, needs newline at the end

**Review by BastiaanOlij — APPROVED** [review]
Looks good, cool addition

File: addons/godot-xr-tools/effects/fade.gdshader (hunk 1 above)
Comment (Malcolmnixon) [reply addons/godot-xr-tools/effects/fade.gdshader:7]: This shader works on the current options (tested on PC, not native):
- Godot 4.1.3 Vulkan & OpenGL
- Godot 4.2.1 Vulkan & OpenGL
- Godot 4.3-dev5 Vulkan & OpenGL

Back in 4.0 we were looking for an option that would satisfy both OpenGL and Vulkan, and I thought the ``cull_disabled` was all we needed.

File: addons/godot-xr-tools/effects/fade.gdshader (hunk 2 above)
Comment (Malcolmnixon) [reply addons/godot-xr-tools/effects/fade.gdshader]: Good catch - fixed.

File: addons/godot-xr-tools/effects/fade.gdshader (hunk 1 above)
Comment (Malcolmnixon) [reply addons/godot-xr-tools/effects/fade.gdshader:7]: What would we need to do if we want this shader to work in Vulkan & OpenGL on Godot 4.1 - 4.3?

In the Reverse Z Progress Report announcement there didn't seem to be anything about how to handle forwards/backwards compatibility.

File: addons/godot-xr-tools/effects/fade.gd [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	var fade := Color(1, 1, 1, 0)
+	var show := false
+	for whom in _faders:
+		var color := _faders[whom] as Color
+		fade = fade.blend(color)
+		show = true
+
+	# Set the shader and show if necessary
+	_material.set_shader_parameter("albedo", fade)
+	_mesh.visible = show
```
Comment (BastiaanOlij) [inline addons/godot-xr-tools/effects/fade.gd:56]: You'll want to add an extra check here if alpha = 0. If the fade is invisible but you don't make the node invisible, you're drawing a full screen rect that is blending with a fully transparent colour. That is unnecessary overhead.

Comment (BastiaanOlij) [reply addons/godot-xr-tools/effects/fade.gd:56]: Never mind, just noticed the alpha check below :P

**Review by BastiaanOlij — APPROVED** [review]
LGTM

**Pull request merged**